pkg/querier/vcs/commit.go

```diff
@@ -0,0 +1,83 @@
+package vcs
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	vcsv1 "github.com/grafana/pyroscope/api/gen/proto/go/vcs/v1"
+)
+
+type gitHubCommitGetter interface {
+	GetCommit(context.Context, string, string, string) (*vcsv1.CommitInfo, error)
+}
+
+// getCommits fetches multiple commits in parallel for a given repository.
+// It attempts to retrieve commits for all provided refs and returns:
+// 1. Successfully fetched commits
+// 2. Errors for failed fetches
+// 3. An overall error if no commits were successfully fetched
+// This function provides partial success behavior, returning any commits
+// that were successfully fetched along with errors for those that failed.
+func getCommits(ctx context.Context, client gitHubCommitGetter, owner, repo string, refs []string) ([]*vcsv1.CommitInfo, []error, error) {
+	type result struct {
+		commit *vcsv1.CommitInfo
+		err    error
+	}
+
+	var wg sync.WaitGroup
```

Comment:
I think ideally we'd use [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup) here so we can set a limit on the number of concurrent requests (maybe like 10).

errgroup also hides away a lot of the channel nonsense, which really limit the number of "read from closed channel" and other similar mistakes (not saying any were made here, but it just takes one careless update!).

Reply:
Nice suggestion to leverage `errgroup` to limit concurrent requests. However, I think we should keep our custom channel error handling to return all errors, since `errgroup` only returns the first error encountered.